File: app/src/main/java/com/sayler/gina/adapter/BaseRecyclerViewAdapter.java
package com.sayler.gina.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;

import java.util.List;


public abstract class BaseRecyclerViewAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

  protected OnItemClickListener<T> onItemClickListener;
  protected List<T> items;
  protected LayoutInflater inflater;
  protected Context context;

  public BaseRecyclerViewAdapter(final Context context, final List<T> items) {
    this.context = context;
    this.items = items;
    this.inflater = LayoutInflater.from(context);
  }



  public void setItems(List<T> items) {
    this.items = items;
    notifyDataSetChanged();
  }

  @Override
  public long getItemId(final int position) {
    return position;
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  private void onItemClicked(final T item, final int position) {
    if (onItemClickListener != null && isItemClickable(1)) {
      onItemClickListener.onItemClicked(item, position);
    }
  }

  public Context getContext() {
    return context;
  }

  protected boolean isItemClickable(int position) {
    return true;
  }

  public void setOnItemClickListener(final OnItemClickListener<T> onItemClickListener) {
    this.onItemClickListener = onItemClickListener;
  }

  public interface OnItemClickListener<T> {
    void onItemClicked(T item, int position);
  }

  public static class RecyclerViewHolderWithOnItemClick<S> extends RecyclerView.ViewHolder implements View.OnClickListener {
    private S item;
    private int position;
    private BaseRecyclerViewAdapter<S> adapter;

    public RecyclerViewHolderWithOnItemClick(final View view, final BaseRecyclerViewAdapter<S> adapter) {
      super(view);
      this.adapter = adapter;
      this.position = -1;
      view.setOnClickListener(this);
    }

    public void setItem(final S item, final int position) {
      this.item = item;
      this.position = position;
    }

    @Override
    public void onClick(final View v) {
      adapter.onItemClicked(item, position);
    }
  }
}

File: app/src/main/java/com/sayler/gina/adapter/BaseRecyclerViewAdapter.kt
package com.sayler.gina.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View

abstract class BaseRecyclerViewAdapter<T>(var context: Context, var items: List<T>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    protected var inflater: LayoutInflater
    var onClickListener: ((item: T, view: View, position: Int) -> Unit)? = null

    init {
        this.context = context
        this.inflater = LayoutInflater.from(context)
    }

    fun setOnClick(onClick: (item: T, view: View, position: Int) -> Unit) {
        onClickListener = onClick
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemCount(): Int {
        return items.size
    }

    private fun onItemClicked(item: T, view: View, position: Int) {
        if (isItemClickable(position)) onClickListener?.invoke(item, view, position)
    }

    private fun isItemClickable(position: Int): Boolean {
        return true
    }

    open class RecyclerViewHolderWithOnItemClick<in S>(view: View, private val adapter: BaseRecyclerViewAdapter<S>) : RecyclerView.ViewHolder(view), View.OnClickListener {
        private var item: S? = null
        private var pos: Int = 0

        init {
            this.pos = -1
            view.setOnClickListener(this)
        }

        fun setItem(item: S, position: Int) {
            this.item = item
            this.pos = position
        }

        override fun onClick(v: View) {
            if (item != null && item is S?) {
                adapter.onItemClicked(item as S, v, pos)
            }
        }
    }
}
